File: hw/Vaca/u3/HW30ObserverAndStrategy/HW30ObserverPatternJava/src/main/java/ec/edu/espe/model/Stock.java

package ec.edu.espe.model;

import java.util.ArrayList;
import java.util.List;
/**
 *
 * @author MSI-PULSE
 */
public class Stock implements Subject {
    private String symbol;
    private double price;
    private List<Observer> observers;

    public Stock(String symbol, double price) {
        this.symbol = symbol;
        this.price = price;
        this.observers = new ArrayList<>();
    }

    public void setPrice(double price) {
        this.price = price;
        notifyObservers();
    }

    public String getSymbol() {
        return symbol;
    }

    public double getPrice() {
        return price;
    }

    @Override
    public void registerObserver(Observer observer) {
        observers.add(observer);
    }

    @Override
    public void removeObserver(Observer observer) {
        observers.remove(observer);
    }

    @Override
    public void notifyObservers() {
        for (Observer observer : observers) {
            observer.update(symbol, price);
        }
    }
}

File: hw/Vaca/u3/HW30ObserverAndStrategy/HW30ObserverPatternJavaScript/model/Stock.js
class Stock extends Subject {
    constructor(symbol, price) {
        super();
        this.symbol = symbol;
        this.price = price;
        this.observers = [];
    }

    setPrice(price) {
        this.price = price;
        this.notifyObservers();
    }

    getSymbol() {
        return this.symbol;
    }

    getPrice() {
        return this.price;
    }

    registerObserver(observer) {
        this.observers.push(observer);
    }

    removeObserver(observer) {
        this.observers = this.observers.filter(obs => obs !== observer);
    }

    notifyObservers() {
        this.observers.forEach(observer => observer.update(this.symbol, this.price));
    }
}
